File: 2017/10/solution.rs
use std::fs::read_to_string;

fn main() {
    let line = read_to_string("input").unwrap();
    let mut data: Vec<u32> = line.trim().chars().map(|x| x as u32).collect();
    data.extend([17, 31, 73, 47, 23].iter());

    let mut nums: Vec<u32> = (0..256).collect();
    let idx = |x| (x % 256) as usize;
    let mut current: u32 = 0;
    let mut skip: u32 = 0;
    for _ in 0..64 {
        for len in data.iter() {
            let mut i = current;
            let mut j = current + len - 1;
            while i < j {
                nums.swap(idx(i), idx(j));
                i += 1;
                j -= 1;
            }
            current += len + skip;
            skip += 1;
        }
    }

    let mut s = String::new();
    let mut index = 0;
    while index < nums.len() {
        let mut n = nums[index];
        for i in index + 1..index + 16 {
            n = n ^ nums[i];
        }
        index += 16;
        s.push_str(&format!("{:02x}", n));
    }
    println!("{}", s);
}

File: 2017/10/solution.py
lines = [line.strip() for line in open("input").readlines()]
line = lines[0]
lens = [ord(ch) for ch in line]
lens.extend([17, 31, 73, 47, 23])

N = 256
ROUNDS = 64
nums = list(range(N))


def idx(n):
    return n % N


skip = 0
current = 0
for _ in range(ROUNDS):
    for l in lens:
        i = current
        j = current + l - 1
        while i < j:
            nums[idx(i)], nums[idx(j)] = nums[idx(j)], nums[idx(i)]
            i += 1
            j -= 1
        current += l + skip
        skip += 1

from more_itertools import chunked

s = ""
for chunk in chunked(nums, 16):
    ch = chunk[0]
    for c in chunk[1:]:
        ch = ch ^ c
    s += f"{ch:02x}"

print(len(s), s)
